src/backend/distributed/master/master_repair_shards.c

```diff
@@ -237,45 +350,341 @@ RecreateTableDDLCommandList(Oid relationId)
+		List *colocatedShardDdlList = NIL;
+
+		colocatedShardDdlList = CopyShardCommandList(colocatedShard, sourceNodeName,
+													 sourceNodePort);
+
+		ddlCommandList = list_concat(ddlCommandList, colocatedShardDdlList);
+	}
+
+	HOLD_INTERRUPTS();
+
+	SendCommandListInSingleTransaction(targetNodeName, targetNodePort, ddlCommandList);
+
+	foreach(colocatedShardCell, colocatedShardList)
 	{
-		copySuccessful = true;
+		ShardInterval *colocatedShard = (ShardInterval *) lfirst(colocatedShardCell);
+		uint64 colocatedShardId = colocatedShard->shardId;
+
+		/*
+		 * If we call this function for repair purposes, we should remove old shard
+		 * placement row.
+		 */
+		InsertShardPlacementRow(colocatedShardId, FILE_FINALIZED,
+								ShardLength(colocatedShardId), targetNodeName,
+								targetNodePort);
 	}
 
-	return copySuccessful;
+	RESUME_INTERRUPTS();
+}
+
+
+/*
+ * RepairShardPlacement repairs a shard along with its co-located shards from a source
+ * node to target node. To do this, it first frops shard placement from target node and
+ * recreates it by using CopyShardPlacement. RepairShardPlacement does not make any
+ * checks about state of the shards. It is caller's responsibility to make those checks
+ * if they are necessary.
+ */
+static void
+RepairShardPlacement(int64 shardId, char *sourceNodeName, int32 sourceNodePort,
+					 char *targetNodeName, int32 targetNodePort)
+{
+	ShardInterval *shardInterval = LoadShardInterval(shardId);
+	List *colocatedShardList = ColocatedShardIntervalList(shardInterval);
+	ListCell *colocatedShardCell = NULL;
+
+	HOLD_INTERRUPTS();
+
+	foreach(colocatedShardCell, colocatedShardList)
+	{
+		ShardInterval *colocatedShard = (ShardInterval *) lfirst(colocatedShardCell);
+		uint64 colocatedShardId = colocatedShard->shardId;
+
+		DeleteShardPlacementRow(colocatedShardId, targetNodeName, targetNodePort);
+	}
+
+	/* CopyShardPlacement function copies given shard with its co-located shards */
+	CopyShardPlacement(shardId, sourceNodeName, sourceNodePort, targetNodeName,
+					   targetNodePort);
+
+	RESUME_INTERRUPTS();
+}
+
+
+/*
+ * CopyShardCommandList generates command list to copy the given shard placement
+ * from the source node to the target node.
+ */
+static List *
+CopyShardCommandList(ShardInterval *shardInterval,
+					 char *sourceNodeName, int32 sourceNodePort)
+{
+	char *shardName = ConstructQualifiedShardName(shardInterval);
+	List *ddlCommandList = NIL;
+	ListCell *ddlCommandCell = NULL;
+	List *copyShardToNodeCommandsList = NIL;
+	StringInfo copyShardDataCommand = makeStringInfo();
+
+	ddlCommandList = RecreateTableDDLCommandList(shardInterval->relationId);
+
+	foreach(ddlCommandCell, ddlCommandList)
+	{
+		char *ddlCommand = lfirst(ddlCommandCell);
+		char *escapedDdlCommand = quote_literal_cstr(ddlCommand);
+
+		StringInfo applyDdlCommand = makeStringInfo();
+		appendStringInfo(applyDdlCommand,
+						 WORKER_APPLY_SHARD_DDL_COMMAND_WITHOUT_SCHEMA,
```

Comment:
Yes. That is correct. In MX we also used WORKER_APPLY_SHARD_DDL_COMMAND_WITHOUT_SCHEMA